src/yz_solrq.erl

```diff
@@ -0,0 +1,283 @@
+        false ->
+            gen_server:reply(From, ok),
+            State2
+    end.
+
+%% Enqueue the entry and return updated state.
+enqueue(E, #indexq{queue = Q, queue_len = L} = IndexQ) ->
+    IndexQ#indexq{queue = queue:in(E, Q),
+                  queue_len = L + 1}.
+
+%% Trigger a flush and return state
+flush(Index, IndexQ, State) ->
+    IndexQ2 = request_worker(Index, IndexQ),
+    update_indexq(Index, IndexQ2, State).
+
+%% Return true if queue is over the high water mark
+over_hwm(#state{all_queue_len = L, queue_hwm = HWM}) ->
+    L > HWM.
+
+%% Request aworker to pull the queue
+maybe_request_worker(Index, #indexq{batch_min = Min} = IndexQ) ->
+    maybe_request_worker(Index, Min, IndexQ).
+
+%% Request a worker to pulll the queue with a provided minimum,
+%% as long as one has not already been requested.
+maybe_request_worker(Index, Min, #indexq{pending_helper = false,
+                                         queue_len = L} = IndexQ) when L >= Min ->
+    request_worker(Index, IndexQ);
+maybe_request_worker(_Index, _Min, IndexQ) ->
+    IndexQ.
+
+%% Notify the solrq workers the index is ready to be pulled.
+request_worker(Index, #indexq{pending_helper = false} = IndexQ) ->
+    Hash = erlang:phash2({Index, self()}),
+    yz_solrq_helper:index_ready(Hash, Index, self()),
+    IndexQ#indexq{pending_helper = true};
+request_worker(_Index, IndexQ) ->
+    IndexQ.
+
+
+%% Send a batch of entries, reply to any blocked vnodes and
+%% return updated state
+send_entries(HPid, Index, #state{all_queue_len = AQL} = State) ->
+    IndexQ = get_indexq(Index, State),
+    {Batch, BatchLen, IndexQ2} = get_batch(IndexQ),
+    yz_solrq_helper:index_batch(HPid, Index, Batch),
+    State2 = State#state{all_queue_len = AQL - BatchLen},
+    case IndexQ2#indexq.queue_len of
+        0 ->
+            delete_indexq(Index, State2);
+        _ ->
+            IndexQ3 = maybe_request_worker(Index, IndexQ2), % may be another full batch
+            IndexQ4 = maybe_start_timer(Index, IndexQ3), % if entries left, restart timer if batch not full
+            update_indexq(Index, IndexQ4, State2)
+    end.
+
+%% Get up to batch_max entries and reset the pending worker/timer href
+get_batch(#indexq{queue = Q, queue_len = L, batch_max = Max} = IndexQ) ->
+    {BatchQ, RestQ} = queue:split(min(L,Max), Q),
+    Batch = queue:to_list(BatchQ),
+    BatchLen = length(Batch),
+    IndexQ2 = IndexQ#indexq{queue = RestQ, queue_len = L - BatchLen,
+                            pending_helper = false, href = undefined},
+    {Batch, BatchLen, IndexQ2}.
+
+
+%% Send replies to blocked vnodes if under the high water mark
+%% and return updated state
+maybe_unblock_vnodes(#state{pending_vnodes = []} = State) ->
+    State;
+maybe_unblock_vnodes(#state{pending_vnodes = PendingVnodes} = State) ->
+    case over_hwm(State) of
+        true ->
+            State;
+        _ ->
+            _ = [gen_server:reply(From, ok) || From <- PendingVnodes],
```

Comment:
return unblocked to make tracing/testing easier - return value is currently ignored anyway.